lib/node_modules/@stdlib/random/base/mt19937/lib/factory.js

```diff
@@ -0,0 +1,163 @@
+* limitations under the License.
+*/
+
+'use strict';
+
+// MODULES //
+
+var setReadOnly = require('@stdlib/utils/define-read-only-property');
+var isPositiveInteger = require('@stdlib/assert/is-positive-integer').isPrimitive;
+var INT32_MAX = require('@stdlib/constants/math/int32-max');
+var UINT32_MAX = require('@stdlib/constants/math/uint32-max');
+var randint32 = require('./rand_int32.js');
+
+
+// VARIABLES //
+
+var NORMALIZATION_CONSTANT = (UINT32_MAX) | 0; // asm type annotation
+var MAX_SEED = (INT32_MAX - 1) | 0; // asm type annotation
+var N = 624;
+var M = 397;
+var UPPER_MASK = 0x80000000; /* Most significant w-r bits */
+var LOWER_MASK = 0x7fffffff; /* Least significant r bits */
+var MATRIX_A = 0x9908b0df;
+
+
+// MAIN //
+
+/**
+* Returns a 32-bit Mersenne Twister pseudorandom number generator.
+*
+* @param {PositiveInteger} [seed] - pseudorandom number generator seed
+* @throws {TypeError} must provide a positive integer
+* @throws {RangeError} must provide a positive integer less than the maximum signed 32-bit integer
+* @returns {Function} Mersenne Twister PRNG
+*
+* @example
+* var mt19937 = factory();
+*
+* var v = mt19937();
+* // returns <number>
+*
+* @example
+* // Return a seeded Mersenne Twister PRNG:
+* var mt19937 = factory( 1234 );
+*
+* var v = mt19937();
+* // returns 822569775
+*/
+function factory( seed ) {
+	var state;
+	var mti;
+	var mt = new Array( N );
+	var s;
+
+	if ( arguments.length ) {
+		if ( !isPositiveInteger( seed ) ) {
+			throw new TypeError( 'invalid input argument. Must provide a positive integer. Value: `' + seed + '`.' );
+		}
+		if ( seed > MAX_SEED ) {
+			throw new RangeError( 'invalid input argument. Must provide a positive integer less than the maximum signed 32-bit integer. Value: `' + seed + '`.' );
+		}
+		state = seed | 0; // asm type annotation
+	} else {
+		state = randint32() | 0; // asm type annotation
+	}
+
+	mt[ 0 ] = state >>> 0;
+	for ( mti = 1; mti < N; mti++ ) {
+		s = mt[ mti - 1 ] ^ ( mt[ mti - 1 ] >>> 30 );
+		mt[ mti ] =
+			( ( ( ( (s & 0xffff0000) >>> 16 ) * 1812433253 ) << 16 ) +
+			( (s & 0x0000ffff) * 1812433253) ) + mti;
+		mt[ mti ] >>>= 0;
+	}
+
+	setReadOnly( mt19937, 'NAME', 'mt19937' );
+	setReadOnly( mt19937, 'SEED', state );
+	setReadOnly( mt19937, 'MIN', 1 );
+	setReadOnly( mt19937, 'MAX', UINT32_MAX - 1 );
+	setReadOnly( mt19937, 'normalized', normalized );
+
+	setReadOnly( normalized, 'NAME', mt19937.NAME );
+	setReadOnly( normalized, 'SEED', mt19937.SEED );
+	setReadOnly( normalized, 'MIN', ( mt19937.MIN - 1.0 ) / NORMALIZATION_CONSTANT );
+	setReadOnly( normalized, 'MAX', ( 4294967294.0 ) * ( 1.0 / 4294967295.0 ) );
+
+	return mt19937;
+
+	/**
+	* Generates a pseudorandom integer on the interval \\( [1,2^{31}-1) \\).
+	*
+	* @private
+	* @returns {PositiveInteger} pseudorandom integer
+	*
+	* @example
+	* var v = mt19937();
+	* // returns <number>
+	*/
+	function mt19937() {
+		var mag01 = [ 0x0, MATRIX_A ];
```

Comment:
This matrix can be moved to the top level scope. At no point are its values overwritten, so a new array instance is not necessary for each invocation. Repeated instantiation just creates unnecessary GC pressure.